internal/protoc/downloader.go

```diff
@@ -0,0 +1,329 @@
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+type downloader struct {
+	logger    *zap.Logger
+	cachePath string
+	protocURL string
+	config    settings.Config
+
+	lock sync.RWMutex
+	// the looked-up and verified to exist base path
+	cachedBasePath string
+}
+
+func newDownloader(config settings.Config, options ...DownloaderOption) *downloader {
+	downloader := &downloader{
+		config: config,
+		logger: zap.NewNop(),
+	}
+	for _, option := range options {
+		option(downloader)
+	}
+	if downloader.config.Compile.ProtobufVersion == "" {
+		downloader.config.Compile.ProtobufVersion = vars.DefaultProtocVersion
+	}
+	return downloader
+}
+
+func (d *downloader) Download() (string, error) {
+	d.lock.RLock()
+	cachedBasePath := d.cachedBasePath
+	d.lock.RUnlock()
+	if cachedBasePath != "" {
+		return cachedBasePath, nil
+	}
+	return d.cache()
+}
+
+func (d *downloader) ProtocPath() (string, error) {
+	basePath, err := d.Download()
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(basePath, "bin", "protoc"), nil
+}
+
+func (d *downloader) WellKnownTypesIncludePath() (string, error) {
+	basePath, err := d.Download()
+	if err != nil {
+		return "", err
+	}
+	return filepath.Join(basePath, "include"), nil
+}
+
+func (d *downloader) Delete() error {
+	basePath, err := d.getBasePathNoVersion()
+	if err != nil {
+		return err
+	}
+	d.cachedBasePath = ""
+	d.logger.Debug("deleting", zap.String("path", basePath))
+	return os.RemoveAll(basePath)
+}
+
+func (d *downloader) cache() (string, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	basePath, err := d.getBasePath()
+	if err != nil {
+		return "", err
+	}
+	downloaded, err := d.isDownloaded(basePath)
+	if err != nil {
+		return "", err
+	}
+	if !downloaded {
+		if err := d.download(basePath); err != nil {
+			return "", err
+		}
+		if _, err := d.isDownloaded(basePath); err != nil {
+			return "", err
+		}
+		d.logger.Debug("protobuf downloaded", zap.String("path", basePath))
+	} else {
+		d.logger.Debug("protobuf already downloaded", zap.String("path", basePath))
+	}
+
+	d.cachedBasePath = basePath
+	return basePath, nil
+}
+
+func (d *downloader) isDownloaded(basePath string) (bool, error) {
+	buffer := bytes.NewBuffer(nil)
+	cmd := exec.Command(filepath.Join(basePath, "bin", "protoc"), "--version")
+	cmd.Stdout = buffer
+	if err := cmd.Run(); err != nil {
+		return false, nil
+	}
+	if d.protocURL != "" {
+		// skip version check since we do not know the version
+		return true, nil
+	}
+	output := strings.TrimSpace(buffer.String())
+	d.logger.Debug("output from protoc --version", zap.String("output", output))
+	if output != fmt.Sprintf("libprotoc %s", d.config.Compile.ProtobufVersion) {
+		return false, nil
+	}
+	return true, nil
+}
+
+func (d *downloader) download(basePath string) (retErr error) {
+	return d.downloadInternal(basePath, runtime.GOOS, runtime.GOARCH)
+}
+
+func (d *downloader) downloadInternal(basePath string, goos string, goarch string) (retErr error) {
```

Comment:
If we don't plan to use anything aside from `runtime.GOOS` and `runtime.GOARCH` we can consolidate these functions and remove `downloadInternal`:
```
func (d *downloader) download(basepath string) (retErr error) {
  url, err := d.getProtocURL(runtime.GOOS, runtime.GOARCH)
  ...
}
```

Reply:
We probably want to have this for testing in the future.